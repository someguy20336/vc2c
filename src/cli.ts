import glob from 'glob'
import program from 'commander'
import { convertFile } from './index.js'
import { writeFileInfo } from './file'

function camelize (str: string) {
  return str.replace(/-(\w)/g, (_, c: string) => c ? c.toUpperCase() : '')
}

// eslint-disable-next-line @typescript-eslint/no-explicit-any
function getCmdOptions (cmd: { options: Array<{ long: string }> }) {
  const args: { [key: string]: boolean | string } = {}
  cmd.options.forEach((o: { long: string }) => {
    const key = camelize(o.long.replace(/^--/, ''))

    if (typeof (cmd as unknown as Record<string, string>)[key] !== 'function' && typeof (cmd as unknown as Record<string, string>)[key] !== 'undefined') {
      args[key] = (cmd as unknown as Record<string, string>)[key]
    }
  })
  return args
}

program
  // eslint-disable-next-line @typescript-eslint/no-var-requires
  .version((require('../../package.json') as { version: string }).version)
  .usage('<command> [options]')

program
  .command('convert <filePath>')
  .description('convert vue component file from class to composition api')
  .option('-v, --view', 'Output file content on stdout, and no write file.')
  .option('-r, --root <root>', 'Set root path for calc file absolute path. Default:`process.cwd()`')
  .option('-c, --config <config>', 'Set vc2c config file path. Default: `\'.vc2c.js\'`')
  // eslint-disable-next-line @typescript-eslint/no-misused-promises
  .action(async (filePath: string, cmd) => {
    const cmdOptions = getCmdOptions(cmd)
    const targetFiles = glob.sync(filePath)

    targetFiles.forEach((targetFile) => {
      const { file, result } = convertFile(targetFile, cmdOptions.root as string, cmdOptions.config as string)

      if (!result.success) {
        console.log("Skipping file: " + targetFile);
        console.log(result.error);
        return;
      }

      if (cmdOptions.view) {
        console.log(result.convertedContent)
        return
      }

      writeFileInfo(file, result.convertedContent)
      console.log('Please check the TODO comments on result.')
    })
  })

program.parse(process.argv)
