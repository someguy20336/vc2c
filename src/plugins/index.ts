import type ts from 'typescript'
import { Vc2cOptions } from '../options'
import { ASTConvertPlugins, ASTResult, ASTConverter, ASTResultKind, RunPluginResult } from './types'
import { copySyntheticComments, addTodoComment, convertNodeToASTResult, isNewInternalHook } from '../utils'
import { log } from '../debug'
import { convertObjName } from './vue-class-component/object/ComponentName'
import { convertObjProps } from './vue-class-component/object/Prop'
import { convertObjData } from './vue-class-component/object/Data'
import { convertName, mergeName } from './vue-class-component/ComponentName'
import { convertModel } from './vue-property-decorator/Model'
import { convertProp, mergeProps } from './vue-property-decorator/Prop'
import { convertDomRef } from './vue-property-decorator/DomRef'
import { convertData } from './vue-class-component/Data'
import { convertGetter, convertSetter, mergeComputed } from './vue-class-component/Computed'
import { convertIntervalHook } from './vue-class-component/IntervalHook'
import { convertWatch } from './vue-property-decorator/Watch'
import { convertEmitMethod } from './vue-property-decorator/Emit'
import { convertMethod } from './vue-class-component/Method'
import { removeThisAndSort } from './removeThisAndSort'
import { convertRender } from './vue-class-component/Render'
import { convertInject } from './vue-property-decorator/Inject'
import { convertProvide } from './vue-property-decorator/Provide'

export function getDefaultPlugins (tsModule: typeof ts): ASTConvertPlugins {
  return {
    [tsModule.SyntaxKind.Decorator]: {
      [tsModule.SyntaxKind.PropertyAssignment]: [
        convertObjName,
        convertObjProps
      ],
      [tsModule.SyntaxKind.MethodDeclaration]: [
        convertObjData,
        convertIntervalHook
      ]
    },
    [tsModule.SyntaxKind.Identifier]: [
      convertName
    ],
    [tsModule.SyntaxKind.HeritageClause]: [
      // TODO: extends Minix
    ],
    [tsModule.SyntaxKind.PropertyDeclaration]: [
      convertModel,
      convertProp,
      convertDomRef,
      convertProvide,
      convertInject,
      convertData
    ],
    [tsModule.SyntaxKind.GetAccessor]: [
      convertGetter
    ],
    [tsModule.SyntaxKind.SetAccessor]: [
      convertSetter
    ],
    [tsModule.SyntaxKind.MethodDeclaration]: [
      convertRender,
      convertIntervalHook,
      convertWatch,
      convertEmitMethod,
      convertMethod
    ],
    after: [
      mergeName,
      mergeProps,
      mergeComputed,
      removeThisAndSort
    ]
  }
}

export function getDecoratorArgumentExpr (tsModule: typeof ts, node: ts.Node): ts.ObjectLiteralExpression | undefined {
  if (tsModule.isCallExpression(node)) {
    if (node.arguments.length > 0) {
      return node.arguments[0] as ts.ObjectLiteralExpression
    }
  }

  return undefined
}

export function getASTResults (
  node: ts.ClassDeclaration,
  options: Vc2cOptions,
  program: ts.Program
): ASTResult<ts.Node>[] {
  const tsModule = options.typescript
  const converterPlugins = options.plugins

  let astResults: ASTResult<ts.Node>[] = []
  node.forEachChild((child) => {
    if (tsModule.isDecorator(child)) {
      const objExpr = getDecoratorArgumentExpr(tsModule, child.expression)
      if (objExpr) {
        objExpr.forEachChild((property) => {
          if (property.kind in converterPlugins[tsModule.SyntaxKind.Decorator]) {
            const objConverters = (converterPlugins[tsModule.SyntaxKind.Decorator] as unknown as { [index: number]: Array<ASTConverter<ts.Node>> })[property.kind]
            let converted = false
            for (const converter of objConverters) {
              const result = converter(property, options, program)
              if (result) {
                astResults.push(result)
                converted = true
                break
              }
            }
            if (!converted) {
              astResults.push(convertNodeToASTResult(tsModule, property))
            }
          }
        })
      }
    } else {
      if (child.kind in converterPlugins) {
        const converters = (converterPlugins as unknown as { [index: number]: Array<ASTConverter<ts.Node>> })[child.kind]
        for (const converter of converters) {
          const result = converter(child, options, program)
          if (result) {
            astResults.push(result)
            break
          }
        }
      }
    }
  })

  for (const fn of converterPlugins.after) {
    astResults = fn(astResults, options, program)
  }

  return astResults
}

export function convertASTResultToSetupFn (astResults: ASTResult<ts.Node>[], options: Vc2cOptions): ts.MethodDeclaration {
  const factory = options.typescript.factory;
  const retVars: Set<string> = new Set<string>();
  
  const returnStatement = addTodoComment(
    options.typescript,
    factory.createReturnStatement(
      factory.createObjectLiteralExpression([
        ...astResults
          .filter((el) => el.kind === ASTResultKind.COMPOSITION)
          .reduce((array, el) => {
            for (let attr of el.attributes) {
              // De-duplicate and remove internal hooks
              if (!array.includes(attr) && !isNewInternalHook(attr)) {
                array.push(attr);
              }
            }
            
            return array;
          }, [] as string[])
          .map((el) => factory.createShorthandPropertyAssignment(
            factory.createIdentifier(el),
            undefined
          ))
      ])
    ),
    'Please remove unused return variable',
    false
  )

  return factory.createMethodDeclaration(
    undefined,
    undefined,
    factory.createIdentifier('setup'),
    undefined,
    undefined,
    [
      factory.createParameterDeclaration(
        undefined,
        undefined,
        factory.createIdentifier(options.setupPropsKey),
        undefined,
        undefined,
        undefined
      ),
      factory.createParameterDeclaration(
        undefined,
        undefined,
        factory.createIdentifier(options.setupContextKey),
        undefined,
        undefined,
        undefined
      )
    ],
    undefined,
    factory.createBlock(
      [
        ...astResults
          .filter((el) => el.kind === ASTResultKind.COMPOSITION)
          .map((el) => el.nodes)
          .reduce((array, el) => array.concat(el), []) as ts.Statement[],
        returnStatement
      ],
      true
    )
  )
}

export function convertASTResultToImport (astResults: ASTResult<ts.Node>[], options: Vc2cOptions): ts.ImportDeclaration[] {
  interface Clause { named: Set<string>, default?: string }

  const factory = options.typescript.factory;
  const importMap = new Map<string, Clause>()
  for (const result of astResults) {
    for (const importInfo of result.imports) {
      const key: string = ('external' in importInfo) ? importInfo.external : importInfo.path
      const temp: Clause = importMap.get(key) ?? { named: new Set() }
      if ((!('default' in temp)) && 'default' in importInfo) {
        temp.default = importInfo.default
      }
      for (const named of importInfo.named || []) {
        temp.named.add(named)
      }
      importMap.set(key, temp)
    }
  }

  !importMap.has('vue') && importMap.set('vue', { named: new Set() });
  // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
  const temp = importMap.get('vue')!;
  if (!temp.named.has("defineComponent")) {
    temp.named.add('defineComponent');
    importMap.set('vue', temp);
  }

  return Array.from(importMap).map((el) => {
    const [key, clause] = el
    return factory.createImportDeclaration(
      undefined,
      factory.createImportClause(
        false,  // TODO What is this
        undefined,
        factory.createNamedImports([...clause.named]
          .map((named) => factory.createImportSpecifier(
            false,
            undefined,
            factory.createIdentifier(named)
          ))
        )
      ),
      factory.createStringLiteral(key)
    )
  })
}

export function runPlugins (
  node: ts.ClassDeclaration,
  options: Vc2cOptions,
  program: ts.Program
): RunPluginResult {
  const tsModule = options.typescript;
  const factory = tsModule.factory;
  log('Start Run ASTPlugins')
  const results = getASTResults(node, options, program)
  log('Finished ASTPlugins')

  log('Make setup function')
  const setupFn = convertASTResultToSetupFn(results, options)
  log('Make default export object')
  const defineComponentExpr = factory.createCallExpression(
      factory.createIdentifier('defineComponent'),
      undefined,
      [factory.createObjectLiteralExpression(
        [
          ...results
            .filter((el) => el.kind === ASTResultKind.OBJECT)
            .map((el) => el.nodes)
            .reduce((array, el) => array.concat(el), []) as ts.PropertyAssignment[],
          setupFn
        ],
        true
      )]
    );

  // Select which one based on the modifier of the class
  // Export default
  let exportExpr: ts.Statement;
  if (tsModule.getModifiers(node)?.some(m => m.kind === tsModule.SyntaxKind.DefaultKeyword)) {
    // export default ...
    exportExpr = tsModule.factory.createExportAssignment(
      undefined,
      undefined,
      defineComponentExpr
    );
  } else {
    // export const <name> ...
    exportExpr = factory.createVariableStatement(
      [factory.createToken(tsModule.SyntaxKind.ExportKeyword)],
      factory.createVariableDeclarationList(
        [factory.createVariableDeclaration(
          factory.createIdentifier(node.name!.getText()),
          undefined,
          undefined,
          defineComponentExpr
        )],
        tsModule.NodeFlags.Const
      )
    );
  }
  
  const exportAssignment = copySyntheticComments(
    tsModule,
    exportExpr,
    node
  );

  return {
    astResults: results,
    statement: exportAssignment
  };
}
