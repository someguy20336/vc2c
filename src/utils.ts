import type vueTemplateParser from 'vue-template-compiler'
import type ts from 'typescript'
import { ASTResult, ASTResultKind, ReferenceKind } from './plugins/types'
export function isVueFile (path: string): boolean {
  return path.endsWith('.vue')
}
export function parseVueFile (vueTemplateParserModule: typeof vueTemplateParser, content: string): vueTemplateParser.SFCDescriptor {
  return vueTemplateParserModule.parseComponent(content)
}

export function getClassDeclarationNodes (tsModule: typeof ts, sourceFile: ts.SourceFile): ts.ClassDeclaration[] | undefined {
  const exportStmts = sourceFile.statements.filter(
    st => tsModule.isClassDeclaration(st) && hasComponentDecorator(tsModule, st)
  )
  if (exportStmts.length === 0) {
    return undefined
  }

  return exportStmts as ts.ClassDeclaration[];
}

export function getDecorator(tsModule: typeof ts, node: ts.Node, decName: string): ts.Decorator | undefined {
  return getDecorators(tsModule, node)
    .find((el) => (el.expression as ts.CallExpression).expression.getText() === decName);
}

export function getDecoratorNames (tsModule: typeof ts, node: ts.Node): string[] {
  return getDecorators(tsModule, node).map((el) => {
      if (tsModule.isCallExpression(el.expression)) {
        return el.expression.expression.getText()
      } else {
        return el.expression.getText()
      }
    });
}

export function hasComponentDecorator(tsModule: typeof ts, node: ts.ClassDeclaration): boolean {
  return getDecoratorNames(tsModule, node).includes('Component');
}

export function getDecorators(tsModule: typeof ts, node: ts.Node): readonly ts.Decorator[] {
  return (tsModule.canHaveDecorators(node) ? tsModule.getDecorators(node) : []) ?? [];
}

const $internalHooks = new Map<string, string | false>([
  ['beforeCreate', false],
  ['created', false],
  ['beforeMount', 'onBeforeMount'],
  ['mounted', 'onMounted'],
  ['unmounted', 'onUnmounted'],
  ['beforeUnmount', 'onBeforeUnmount'],
  ['beforeDestroy', 'onBeforeUnmount'],
  ['destroyed', 'onUnmounted'],
  ['beforeUpdate', 'onBeforeUpdate'],
  ['updated', 'onUpdated'],
  ['activated', 'onActivated'],
  ['deactivated', 'onDeactivated'],
  ['render', 'onRender'],
  ['errorCaptured', 'onErrorCaptured'], // 2.5
  ['serverPrefetch', 'onServerPrefetch'] // 2.6
])

export function isInternalHook (methodName: string): boolean {
  return $internalHooks.has(methodName)
}

export function isNewInternalHook(name: string): boolean {
  for (let hook of $internalHooks.values()) {
    if (hook === name) {
      return true;
    }
  }
  
  return false;
}

export function getMappedHook (methodName: string): string | undefined | false {
  return $internalHooks.get(methodName)
}

export function isPrimitiveType (tsModule: typeof ts, returnType: ts.Type): boolean {
  return !!(returnType.flags & tsModule.TypeFlags.NumberLike) ||
    !!(returnType.flags & tsModule.TypeFlags.StringLike) ||
    !!(returnType.flags & tsModule.TypeFlags.BooleanLike) ||
    !!(returnType.flags & tsModule.TypeFlags.Null) ||
    !!(returnType.flags & tsModule.TypeFlags.Undefined)
}

export function copySyntheticComments<T extends ts.Node> (tsModule: typeof ts, node: T, copyNode: ts.Node): T {
  const leadingComments = tsModule.getLeadingCommentRanges(copyNode.getSourceFile().getFullText(), copyNode.pos) || []
  const trailingComments = tsModule.getTrailingCommentRanges(copyNode.getSourceFile().getFullText(), copyNode.end) || []

  const getCommentText = (comment: ts.CommentRange) => {
    return copyNode.getSourceFile().getFullText().slice(comment.pos, comment.end)
      .replace(/\/\//g, '')
      .replace(/\/\*/g, '')
      .replace(/\*\//g, '')
      .replace(/ {2}\* ?/g, '* ')
      .replace(/ \*\//g, '*/')
      .replace(/ {2}$/g, '')
  }

  let result = node
  for (const comment of leadingComments) {
    const text = getCommentText(comment)
    result = tsModule.addSyntheticLeadingComment(result, comment.kind, text, comment.hasTrailingNewLine)
  }

  for (const comment of trailingComments) {
    const text = getCommentText(comment)
    result = tsModule.addSyntheticTrailingComment(result, comment.kind, text, comment.hasTrailingNewLine)
  }

  return node
}

export function removeComments<T extends ts.Node> (tsModule: typeof ts, node: T): T | ts.StringLiteral {
  if (tsModule.isStringLiteral(node)) {
    return tsModule.factory.createStringLiteral(node.text)
  }
  return node
}

export function addTodoComment<T extends ts.Node> (tsModule: typeof ts, node: T, text: string, multiline: boolean): T {
  return tsModule.addSyntheticLeadingComment(
    node,
    (multiline) ? tsModule.SyntaxKind.MultiLineCommentTrivia : tsModule.SyntaxKind.SingleLineCommentTrivia,
    ` TODO: ${text}`
  )
}

export function convertNodeToASTResult<T extends ts.Node> (tsModule: typeof ts, node: T): ASTResult<T> {
  return {
    imports: [],
    kind: ASTResultKind.OBJECT,
    reference: ReferenceKind.NONE,
    attributes: [],
    tag: 'IheritObjProperty',
    nodes: [
      addTodoComment(tsModule, node, 'Can\'t convert this object property.', false)
    ]
  }
}

// ts.createIdentifier() cannot call getText function, it's a hack.
export function createIdentifier (tsModule: typeof ts, text: string): ts.Identifier {
  const temp = tsModule.factory.createIdentifier(text)
  // eslint-disable-next-line @typescript-eslint/unbound-method
  temp.getText = () => text
  return temp
}

const newLineMarker = "//^^^^^NewLineMarker^^^^^";

export function encodeEmptyLines(text: string) {    
    const lines = text.split(/\r?\n/);    
    const commentedLines = lines.map(line => line.trim() == '' ? newLineMarker : line);    
    return commentedLines.join("\r\n");
}

export function decodeEmptyLines(text: string){    
    var lines = text.split(/\r?\n/);
    const uncommentedLines = lines.map(line => line.trim() == newLineMarker ? '' : line);    
    return uncommentedLines.join("\r\n");
}