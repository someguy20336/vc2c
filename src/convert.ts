import type ts from 'typescript'
import { getClassDeclarationNodes, hasComponentDecorator } from './utils'
import { convertASTResultToImport, runPlugins } from './plugins'
import { Vc2cOptions } from './options'
import { log } from './debug'
import { ASTResult, ConvertResult } from './plugins/types'

const vueClassModules = [
  'vue-class-component',
  'vue-property-decorator'
]

export function convertAST (sourceFile: ts.SourceFile, options: Vc2cOptions, program: ts.Program): ConvertResult {
  const tsModule = options.typescript

  log('check vue class library')
  const vueClassModuleImportStatement = sourceFile.statements
    .find((statement) => {
      if (tsModule.isImportDeclaration(statement)) {
        if (vueClassModules.includes((statement.moduleSpecifier as ts.StringLiteral).text)) {
          return true
        }
      }
      return false
    })
  if (!vueClassModuleImportStatement) {
    return {
      success: false,
      convertedContent: "",
      error: "No vue class library in this file"
    }
  }

  log('check default export class')

  const otherStatements = sourceFile.statements
    .map((el) => el)
    .filter((el) =>
      !((tsModule.isClassDeclaration(el) && hasComponentDecorator(tsModule, el)) ||
      (tsModule.isImportDeclaration(el) && vueClassModules.includes((el.moduleSpecifier as ts.StringLiteral).text)) ||
      (tsModule.isImportDeclaration(el) && (el.moduleSpecifier as ts.StringLiteral).text === 'vue'))
    )

  let resultStatements = otherStatements;

  const classNodes = getClassDeclarationNodes(options.typescript, sourceFile)
  if (!classNodes) {
    return {
      success: false,
      convertedContent: "",
      error: "No vue class components found in this file"
    }
  }

  let astResults: ASTResult<ts.Node>[] = [];

  for (let cNode of classNodes) {
    const result = runPlugins(cNode, options, program);
    astResults = astResults.concat(result.astResults);
    resultStatements.push(result.statement);
  }
  
  log('Make ImportDeclaration')
  resultStatements = resultStatements.concat(convertASTResultToImport(astResults, options));

  resultStatements = [
    ...resultStatements.filter((el) => tsModule.isImportDeclaration(el)),
    ...resultStatements.filter((el) => !tsModule.isImportDeclaration(el))
  ]

  log('output result code')
  const printer = tsModule.createPrinter()
  const result = printer.printFile(tsModule.factory.updateSourceFile(sourceFile, resultStatements))

  return {
    success: true,
    convertedContent: result,
    error: ""
  };
}
