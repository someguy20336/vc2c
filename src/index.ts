import { getSingleFileProgram } from './parser'
import { convertAST } from './convert'
import { InputVc2cOptions, getDefaultVc2cOptions, mergeVc2cOptions } from './options'
import { format } from './format'
import path from 'path'
import { readVueSFCOrTsFile, existsFileSync, FileInfo } from './file'
import { setDebugMode } from './debug'
import * as BuiltInPlugins from './plugins/builtIn'
import { ConvertResult } from './plugins/types'
import { decodeEmptyLines, encodeEmptyLines } from './utils'

export function convert (content: string, inputOptions: InputVc2cOptions): ConvertResult {
  const options = mergeVc2cOptions(getDefaultVc2cOptions(inputOptions.typescript), inputOptions);
  const { ast, program } = getSingleFileProgram(encodeEmptyLines(content), options)

  const convResult = convertAST(ast, options, program);
  if (convResult.success) {
    convResult.convertedContent = format(decodeEmptyLines(convResult.convertedContent), options);
  }

  return convResult;
}

export function convertFile (filePath: string, root: string, config: string): { file: FileInfo, result: ConvertResult } {
  root = (typeof root === 'string')
    ? (
      path.isAbsolute(root) ? root : path.resolve(process.cwd(), root)
    )
    : process.cwd()
  config = (typeof config === 'string') ? config : '.vc2c.js'
  if (config.endsWith('.ts')) {
    require('ts-node/register')
  }
  const inputOptions: InputVc2cOptions = config && existsFileSync(path.resolve(root, config))
    // eslint-disable-next-line @typescript-eslint/no-var-requires
    ? require(path.resolve(root, config)) as InputVc2cOptions
    : {}
  const options = mergeVc2cOptions(getDefaultVc2cOptions(inputOptions.typescript), inputOptions)
  options.root = root

  if (options.debug) {
    setDebugMode(true)
  }

  const file = readVueSFCOrTsFile(filePath, options)
  return {
    file,
    result: convert(file.content, options)
  }
}

export * from './plugins/types'
export { BuiltInPlugins }
export * from './utils'
export { getDefaultVc2cOptions, Vc2cOptions } from './options'
