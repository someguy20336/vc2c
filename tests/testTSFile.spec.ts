import { convertFile } from '../src'
import { FileKind } from '../src/file'
import path from 'path'
import util from 'util'
import { exec } from 'child_process'

const execAsync = util.promisify(exec)

describe('testTSFile', () => {
  const defComponentPath = 'fixture/Input.ts';
  const commonCompPath = 'fixture/CommonUseWithWhitespaceTest.ts';
  const multComponentPath = "fixture/MultipleComponents.ts";
  const multComponentDefaultPath = "fixture/MultipleComponentsWithDefault.ts";
  const componentAndNonCompPath = "fixture/ComponentWithNonComponentClass.ts";

  it('converts a file with a single default component', () => {
    const { file, result } = convertFile(defComponentPath, __dirname, '')
    expect(file.fsPath.includes(path.basename(defComponentPath))).toBeTruthy()
    expect(path.isAbsolute(file.fsPath)).toBeTruthy()
    expect(file.kind).toBe(FileKind.TS)
    expect(file).not.toHaveProperty('start')
    expect(file).not.toHaveProperty('end')
    expect(result.convertedContent).toMatchSnapshot()
  });

  it('converts a file with a single default component with vertical whitespace', () => {
    const { file, result } = convertFile(commonCompPath, __dirname, '')
    expect(file.fsPath.includes(path.basename(commonCompPath))).toBeTruthy()
    expect(path.isAbsolute(file.fsPath)).toBeTruthy()
    expect(file.kind).toBe(FileKind.TS)
    expect(file).not.toHaveProperty('start')
    expect(file).not.toHaveProperty('end')
    expect(result.convertedContent).toMatchSnapshot()
  });

  it('converts a file with multiple components', () => {
    const { file, result } = convertFile(multComponentPath, __dirname, '')
    expect(file.fsPath.includes(path.basename(multComponentPath))).toBeTruthy()
    expect(path.isAbsolute(file.fsPath)).toBeTruthy()
    expect(file.kind).toBe(FileKind.TS)
    expect(file).not.toHaveProperty('start')
    expect(file).not.toHaveProperty('end')
    expect(result.convertedContent).toMatchSnapshot()
  });

  it('converts a file with multiple components and one default', () => {
    const { file, result } = convertFile(multComponentDefaultPath, __dirname, '')
    expect(file.fsPath.includes(path.basename(multComponentDefaultPath))).toBeTruthy()
    expect(path.isAbsolute(file.fsPath)).toBeTruthy()
    expect(file.kind).toBe(FileKind.TS)
    expect(file).not.toHaveProperty('start')
    expect(file).not.toHaveProperty('end')
    expect(result.convertedContent).toMatchSnapshot()
  });

  it('converts a file with a component and non-component class', () => {
    const { file, result } = convertFile(componentAndNonCompPath, __dirname, '')
    expect(file.fsPath.includes(path.basename(componentAndNonCompPath))).toBeTruthy()
    expect(path.isAbsolute(file.fsPath)).toBeTruthy()
    expect(file.kind).toBe(FileKind.TS)
    expect(file).not.toHaveProperty('start')
    expect(file).not.toHaveProperty('end')
    expect(result.convertedContent).toMatchSnapshot()
  });
})
