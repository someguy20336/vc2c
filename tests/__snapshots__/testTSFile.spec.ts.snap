// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`testTSFile converts a file with a component and non-component class 1`] = `
"import { defineComponent } from 'vue'

export class NonComponent {
  public testOne: string = '1'
  public testTwo: number = 2
}

/**
 * My basic tag
 */
export const BasicPropertyClass = defineComponent({
  props: {
    /**
     * My foo
     */
    foo: { type: Number, default: false }
  },
  name: 'BasicPropertyClass',
  setup(props, context) {
    /**
     * My greeting
     */
    const hello = () => {
      console.log('eh')
    }

    // TODO: Please remove unused return variable
    return { hello }
  }
})
"
`;

exports[`testTSFile converts a file with a single default component 1`] = `
"import {
  computed,
  reactive,
  toRefs,
  ref,
  onBeforeUnmount,
  onUnmounted,
  watch,
  onMounted,
  defineComponent
} from 'vue'

/**
 * My basic tag
 */
export default defineComponent({
  model: {
    prop: 'checked',
    event: 'change'
  },
  props: {
    bar: String,
    checked: { type: Boolean },
    /**
     * My foo
     */
    foo: { type: Number, default: false }
  },
  name: 'oao',
  setup(props, context) {
    /**
     * My count
     */
    const count = computed(() => {
      return store.state.count
    })

    const { a } = (() => {
      const a = 'pa'

      return toRefs(
        reactive({
          a: a
        })
      )
    })()

    const anotherComponent = ref<HTMLElement>(null)

    /**
     * My msg
     */
    const msg = ref('Vetur means \\"Winter\\" in icelandic.') // foo

    onBeforeUnmount(() => {
      context.emit('tearing down')
    })

    onUnmounted(() => {
      console.log('destroyed')
    })

    const click = (c: string) => {
      context.emit('click', c)
    }

    const value = computed({
      get: () => {
        return props.checked
      },
      set: (value: boolean) => {
        context.emit('change', value)
      }
    })

    watch(
      props.checked,
      (val: boolean, newVal: boolean) => {
        console.log(val, newVal)
        console.log(anotherComponent.value!)
      },
      { deep: true, immediate: true }
    )
    watch(msg, (val: string, newVal: string) => {
      console.log(val, newVal)
    })

    onMounted(() => {
      click('oao')
    })

    const removeItem = () => {
      if (props.foo > 10) {
        context.emit('remove-item', 10)
      } else {
        context.emit('remove-item', props.foo)
      }
    }

    /**
     * My greeting
     */
    const hello = () => {
      console.log(msg.value)
    }

    removeItem()
    // TODO: Please remove unused return variable
    return {
      count,
      a,
      anotherComponent,
      msg,
      click,
      value,
      checked,
      removeItem,
      hello
    }
  }
})
"
`;

exports[`testTSFile converts a file with a single default component with vertical whitespace 1`] = `
"import {
  ref,
  onBeforeUnmount,
  onUnmounted,
  computed,
  watch,
  onMounted,
  defineComponent
} from 'vue'

/**
 * My basic tag
 */
export default defineComponent({
  props: {
    /**
     * My foo
     */
    aProperty: { type: Number, default: false },
    propWithNothing: {}
  },
  name: 'general-usage',
  template: '#test-template',
  setup(props, context) {
    /**
     * My msg
     */
    const msg = ref('Vetur means \\"Winter\\" in icelandic.')

    onBeforeUnmount(() => {
      context.emit('tearing down')
    })

    onUnmounted(() => {
      console.log('destroyed')

      console.log('destroyed')
      console.log('destroyed')


      console.log('destroyed')
      console.log('destroyed')
    })

    /**
     * Computed
     */
    const computedGetterOnly = computed(() => {
      return props.aProperty + 1
    })

    const computedGetterSetter = computed({
      get: () => {
        return props.propWithNothing + '-appended'
      },
      set: (value: string) => {
        context.emit('change', value)
      }
    })

    watch(msg, (val: string, oldVal: string) => {
      console.log(val, oldVal)
      console.log(val, oldVal)

      console.log(val, oldVal)
      console.log(val, oldVal)
    })

    /**
     * My greeting
     */
    const hello = () => {
      console.log(msg.value)
    }

    watch(
      computedGetterOnly,
      (val: number, oldVal: number | undefined) => {
        console.log(val, oldVal)
      },
      { deep: true, immediate: true }
    )

    onMounted(() => {
      msg.value = 'Changed'

      const a = 2
      const b = 5

      const c = a + b

      hello()
    })
    hello()

    hello()
    console.log('created')
    // TODO: Please remove unused return variable
    return { msg, computedGetterOnly, computedGetterSetter, hello }
  }
})
"
`;

exports[`testTSFile converts a file with multiple components 1`] = `
"import { defineComponent } from 'vue'

/**
 * My basic tag
 */
export const BasicPropertyClass = defineComponent({
  props: {
    /**
     * My foo
     */
    foo: { type: Number, default: false }
  },
  name: 'BasicPropertyClass',
  setup(props, context) {
    /**
     * My greeting
     */
    const hello = () => {
      console.log('eh')
    }

    // TODO: Please remove unused return variable
    return { hello }
  }
})

export const AnotherComponent = defineComponent({
  props: {
    /**
     * My foo
     */
    foo: { type: Number, default: false }
  },
  name: 'AnotherComponent',
  setup(props, context) {
    /**
     * My greeting
     */
    const hello2 = () => {
      console.log('eh')
    }

    // TODO: Please remove unused return variable
    return { hello2 }
  }
})

export const AThirdComponent = defineComponent({
  props: {
    /**
     * My foo
     */
    foo: { type: Number, default: false }
  },
  name: 'AThirdComponent',
  setup(props, context) {
    /**
     * My greeting
     */
    const hello3 = () => {
      console.log('eh')
    }

    // TODO: Please remove unused return variable
    return { hello3 }
  }
})
"
`;

exports[`testTSFile converts a file with multiple components and one default 1`] = `
"import { defineComponent } from 'vue'

/**
 * My basic tag
 */
export const BasicPropertyClass = defineComponent({
  props: {
    /**
     * My foo
     */
    foo: { type: Number, default: false }
  },
  name: 'BasicPropertyClass',
  setup(props, context) {
    /**
     * My greeting
     */
    const hello = () => {
      console.log('eh')
    }

    // TODO: Please remove unused return variable
    return { hello }
  }
})

export const AnotherComponent = defineComponent({
  props: {
    /**
     * My foo
     */
    foo: { type: Number, default: false }
  },
  name: 'AnotherComponent',
  setup(props, context) {
    /**
     * My greeting
     */
    const hello2 = () => {
      console.log('eh')
    }

    // TODO: Please remove unused return variable
    return { hello2 }
  }
})

export default defineComponent({
  props: {
    /**
     * My foo
     */
    foo: { type: Number, default: false }
  },
  name: 'AThirdDefaultComponent',
  setup(props, context) {
    /**
     * My greeting
     */
    const hello3 = () => {
      console.log('eh')
    }

    // TODO: Please remove unused return variable
    return { hello3 }
  }
})
"
`;
